src/ripple/protocol/impl/SecretKey.cpp

```diff
@@ -54,45 +53,160 @@ SecretKey::to_string() const
+        for (std::uint32_t subseq = 0; subseq != 128; ++subseq)
+        {
+            copy_uint32(buf.data() + 37, subseq);
+
+            auto const ret = sha512Half_s(buf);
+
+            if (secp256k1_ec_seckey_verify(secp256k1Context(), ret.data()) == 1)
+            {
+                secure_erase(buf.data(), buf.size());
+                return ret;
+            }
+        }
+
+        Throw<std::runtime_error>("Unable to derive generator from seed");
+    }
 
 public:
     explicit Generator(Seed const& seed)
+        : root_(deriveDeterministicRootKey(seed))
     {
-        // FIXME: Avoid copying the seed into a uint128 key only to have
-        //        generateRootDeterministicPublicKey copy out of it.
-        uint128 ui;
-        std::memcpy(ui.data(), seed.data(), seed.size());
-        gen_ = generateRootDeterministicPublicKey(ui);
+        secp256k1_pubkey pubkey;
+        if (secp256k1_ec_pubkey_create(
+                secp256k1Context(), &pubkey, root_.data()) != 1)
+            LogicError("derivePublicKey: secp256k1_ec_pubkey_create failed");
+
+        auto len = generator_.size();
+
+        if (secp256k1_ec_pubkey_serialize(
+                secp256k1Context(),
+                generator_.data(),
+                &len,
+                &pubkey,
+                SECP256K1_EC_COMPRESSED) != 1)
+            LogicError("derivePublicKey: secp256k1_ec_pubkey_serialize failed");
     }
 
-    /** Generate the nth key pair.
+    ~Generator()
+    {
+        secure_erase(root_.data(), root_.size());
+        secure_erase(generator_.data(), generator_.size());
+    }
 
-        The seed is required to produce the private key.
-    */
+    /** Generate the nth key pair. */
     std::pair<PublicKey, SecretKey>
-    operator()(Seed const& seed, std::size_t ordinal) const
+    operator()(std::size_t ordinal) const
     {
-        // FIXME: Avoid copying the seed into a uint128 key only to have
-        //        generatePrivateDeterministicKey copy out of it.
-        uint128 ui;
-        std::memcpy(ui.data(), seed.data(), seed.size());
-        auto gsk = generatePrivateDeterministicKey(gen_, ui, ordinal);
-        auto gpk = generatePublicDeterministicKey(gen_, ordinal);
-        SecretKey const sk(Slice{gsk.data(), gsk.size()});
-        PublicKey const pk(Slice{gpk.data(), gpk.size()});
-        secure_erase(ui.data(), ui.size());
-        secure_erase(gsk.data(), gsk.size());
-        return {pk, sk};
+        // Generates Nth secret key:
+        auto gsk = [this, tweak = calculateTweak(ordinal)]() {
+            auto rpk = root_;
+
+            if (secp256k1_ec_privkey_tweak_add(
+                    secp256k1Context(), rpk.data(), tweak.data()) == 1)
+            {
+                SecretKey sk{Slice{rpk.data(), rpk.size()}};
+                secure_erase(rpk.data(), rpk.size());
+                return sk;
+            }
+
+            LogicError("Unable to add a tweak!");
+        }();
+
```

Comment:
The old code that the ability to generate public keys without the private key. I believe this was never used, so we may as well lose the functionality. (but again, just pointing out a difference between the old and new implementation).

Reply:
Correct: the functionality was never used. It may be worth implementing that mechanism too for reference, but I hesitate to write code (especially cryptographic code) that isn't used. And if I'm being honest, I'd rather remove as much of this code from `rippled` as possible.

This kind of thing should be handled by a separate tool that can generate/derive keys.